File: Graph(그래프)/백준_10451_순열사이클/순열 사이클.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] arr = new int[1001];
    static boolean[] check = new boolean[1001];

    static int cycle(int n) {
        int count = 0;
        for (int i = 1; i <= n; i++) {
            if (check[i] == false) {
                count++;
                int start = i;
                int end = arr[i];
                while (start != end) {
                    check[end] = true;
                    end = arr[end];
                }
            }
        }
        return count;
    }

    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        int T = s.nextInt();

        while(T-->0) {
            int n = s.nextInt();
            for (int i = 1; i <= n; i++) {
                arr[i] = s.nextInt();
                check[i] = false;
            }
            System.out.println(cycle(n));
        }
    }
}

File: Graph(그래프)/백준_10451_순열사이클/순열 사이클.kt
import java.util.*
import kotlin.math.*

var arr = IntArray(1001, {0})
var check = BooleanArray(1001, {false})

fun cycle(n : Int): Int{
    var count = 0
    for(i in 1..n){
        if(check[i] == false){
            count++
            var start = i
            var end = arr[i]
            while(start != end){
                check[end] = true
                end=  arr[end]
            }
        }
    }
    return count
}

fun main(args: Array<String>){
    val s = Scanner(System.`in`)
    var T = s.nextInt()
    while(T-- > 0){
        val n = s.nextInt()
        for(i in 1..n){
            arr[i] = s.nextInt()
            check[i] = false
        }
        println(cycle(n))
    }
}